tests/device_tests/test_protection_levels.py

```diff
@@ -32,239 +33,362 @@
+def test_apply_settings(client):
+    with client:
+        assert client.features.pin_protection is True
+        client.use_pin_sequence([PIN4])
+        client.set_expected_responses(
+            [
+                _pin_request(client),
+                messages.ButtonRequest,
+                messages.Success,
+                messages.Features,
+            ]
+        )  # TrezorClient reinitializes device
+        device.apply_settings(client, label="nazdar")
+        assert client.features.passphrase_protection is True
+
+
+def test_change_pin(client):
+    with client:
+        assert client.features.pin_protection is True
+        # Models differ both in PIN sequence (TT requires one more input)
+        # and in expected responses (TT has more of them)
+        if client.features.model == "1":
             client.use_pin_sequence([PIN4, PIN4, PIN4])
-            client.set_expected_responses(
-                [
-                    messages.ButtonRequest,
-                    messages.PinMatrixRequest,
-                    messages.PinMatrixRequest,
-                    messages.PinMatrixRequest,
-                    messages.Success,
-                    messages.Features,
-                ]
-            )
-            device.change_pin(client)
-
-    def test_ping(self, client):
-        with client:
-            client.set_expected_responses([messages.ButtonRequest, messages.Success])
-            client.ping("msg", True)
-
-    @pytest.mark.setup_client(pin=PIN4, passphrase=True)
-    def test_get_entropy(self, client):
-        with client:
-            client.set_expected_responses([messages.ButtonRequest, messages.Entropy])
-            misc.get_entropy(client, 10)
-
-    @pytest.mark.setup_client(pin=PIN4, passphrase=True)
-    def test_get_public_key(self, client):
-        with client:
+            expected_responses = [
+                messages.ButtonRequest,
+                messages.PinMatrixRequest,
+                messages.PinMatrixRequest,
+                messages.PinMatrixRequest,
+                messages.Success,
+                messages.Features,
+            ]
+        else:
+            client.use_pin_sequence([PIN4, PIN4, PIN4, PIN4])
+            expected_responses = [
+                messages.ButtonRequest,
+                messages.ButtonRequest,
+                messages.ButtonRequest,
+                messages.ButtonRequest,
+                messages.ButtonRequest,
+                messages.ButtonRequest,
+                messages.Success,
+                messages.Features,
+            ]
+        client.set_expected_responses(expected_responses)
+        device.change_pin(client)
+
+    assert client.features.passphrase_protection is True
+
+
+@pytest.mark.setup_client(pin=None, passphrase=False)
+def test_ping(client):
+    with client:
+        assert client.features.pin_protection is False
+        assert client.features.passphrase_protection is False
```

Comment:
we usually put these asserts before the with-block, and do only the things the with-block needs inside (so expected responses, pin sequences, passphrases, and the call (or calls) to Trezor)

Reply:
Yes, I like the idea of the `_assert_protection()` function, so I will call it before the `with` block in all the tests, so after these assertions there are only test-relevant things

Reply:
Fixed in 699ae28